[VectorCombine] foldShuffleOfBinops - failure to track OperandValueInfo

KavinTheG opened the pull request:
Resolves #170500.

Implemented mergeInfo static helper to return common TTI::OperandValueInfo data . 

Added common OperandValueInfo `Op0Info` && `Op1Info` to NewCost calculation.

KavinTheG committed:
[VectorCombine] foldShuffleOfBinops - failure to track OperandValueInfo
data in getArithmeticInstrCost calls #170500

Add OperandValueInfo to NewCost calculation in foldShuffleOfBinops and
add static helper method mergeInfo in TargetTransformInfo to merge
common TTI::OperandValueInfo data.

KavinTheG committed:
Add mergeInfo variant to accpet OperandValueInfo inputs.
Add more test cases as well.

KavinTheG commented:
- Added an additional variant of mergeInfo to take Value args 
```C++
  LLVM_ABI static OperandValueInfo mergeInfo(const Value *X, const Value *Y);
  LLVM_ABI static OperandValueInfo mergeInfo(const OperandValueInfo X, const OperandValueInfo Y, bool IsEqual);
```

- I noticed that `foldShuffleOfBinops` already had an X == Z & Y == W statements, so I called `getOperandInfo(X)` if they're equal, and     `mergeInfo(TTI::getOperandInfo(X), TTI::getOperandInfo(Z), false)` otherwise. Ultimately makes mergeInfo with value inputs *not* needed for this patch, so I can remove it in the next commit if it's unnecessary fluff. Thought to keep it in to make it easier to use in other scenarios 

- Added the following test cases after examining `X86TTImpl::getArithmeticInstrCost`
	- Uniform constant shift 
	- Non uniform constant shift 
	- Uniform constant & pow2 mul
	- Nonuniform constant mul (non pow2)
	- Uniform constant & non pow2 mul 
	- Uniform mul (to test OK_UniformValue)
	- Uniform constant mul (different vector size)
	- Uniform constant add & sub
	- Uniform constant AND

KavinTheG commented:
Ping.

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -2437,21 +2437,28 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
+    Op1Info = TTI.getOperandInfo(Y);
+  } else
+    Op1Info =
+        TTI.mergeInfo(TTI::getOperandInfo(Y), TTI::getOperandInfo(W), false);
```
`TTI::OperandValueInfo Op1Info = TTI.mergeInfo(Y, W);`

RKSimon commented on llvm/test/Transforms/VectorCombine/X86/shuffle-of-binops.ll:
```diff
@@ -4,11 +4,172 @@
 
 declare void @use(<4 x i32>)
 
+; Uniform constant shift
```
Are there no tests that have a different cost benefit for SSE2 vs AVX2?

KavinTheG replied:
Apologies for the late update. For clarification, should I only be adding test cases where both SSE2 and AVX2 cost table has an entry with  the same vector type and instruction? Like below?

```C++
  static const CostKindTblEntry AVX2UniformCostTable[] = {
    // Uniform splats are cheaper for the following instructions.
    { ISD::SHL,  MVT::v16i8,  { 3, 5, 5, 7 } }, // psllw + pand.
    ...
  }
  static const CostKindTblEntry SSE2UniformCostTable[] = {
    // Uniform splats are cheaper for the following instructions.
    { ISD::SHL,  MVT::v16i8, {  9, 10, 6, 9 } }, // psllw + pand.
  }
```

RKSimon commented on llvm/test/Transforms/VectorCombine/X86/shuffle-of-binops.ll:
```diff
@@ -4,11 +4,172 @@
+  ret <8 x i32> %res
+}
+
+; Uniform constant ADD
```
only shift/mul tests are likely to be useful

RKSimon reviewed (changes requested):
Feedback still needs to be addressed

KavinTheG committed:
Adds and removes tests to better compare AVX2 vs SSE2 costs. Refactors
mergeInfo to OperandValueInfo member function mergeWith and refactors
variant to commonOperandInfo.

KavinTheG commented:
- Refactored `mergeInfo(const OperandValueInfo X, const OperandValueInfo Y, bool IsEqual)` to  an OperandValueInfo member function `mergeWith(const OperandValueInfo Y)`
- Refactored variant `mergeInfo` to `commonOperandInfo` 
- Removed non-shift & non-mul tests 
- Added 
  - Variable shifts (Non-constant & non-uniform)
  - Different vector sizes
 
Perhaps I am missing something, but from what I have tried so far, only variable shifts and OK_UniformValue tests (where a value to be splat is given) produces different IR.

RKSimon commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1254,6 +1255,19 @@ class TargetTransformInfo {
+
+      MergeProp = Properties == OpInfoY.Properties ? Properties : OP_None;
+      return {MergeKind, MergeProp};
+    }
```
simplify:
```
    OperandValueInfo mergeWith(const OperandValueInfo OpInfoY) {
      OperandValueKind MergeKind = OK_AnyValue;
      if (isConstant() && OpInfoY.isConstant())
        MergeKind = OK_NonUniformConstantValue;
    
      OperandValueProperties MergeProp = OP_None;
	  if (Properties == OpInfoY.Properties)
	    MergeProp = Properties;
		
      return {MergeKind, MergeProp};
    }
```

RKSimon reviewed (changes requested):
a few minor simplifications

KavinTheG committed:
Implemented suggest changes.

KavinTheG commented:
- Implemented suggest changes:
  - `Op0Info = TTI.commonOperandInfo(X, Z)` & `Op1Info` above their respective if statements
    - Removed `getOperandInfo`  call within if statements (X == Z and Y == W) as this is already performed within `commonOperandInfo`
  - Simplified `mergeWith`
  - Moved new test cases to the bottom

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into vectorcombine-fix-binop-shuffle-ovi
